### ISSUE TITLE
State of Art ：Think about the Next-Gen waifu2x 

### ISSUE THREAD

**Issue by 2ji3150** [nagadomi/waifu2x#236]
https://github.com/LimBee/NTIRE2017
This is the best algorithm "EDSR" in the NITRE2017 and it is implemented on tourch. If I am not wrong it is the same framework with the waifu2x.
Also a pytourch version in on github.
https://github.com/thstkdgus35/EDSR-PyTorch

The paper is here：
https://arxiv.org/pdf/1707.02921.pdf

I realize that waifu2x is aim for web service. But each of us dearest wish to what is the next level of the Anime-Style Art Super-Resolution would be. 
If the quality becomes higher and higher it should do help the 2-d anime ceators.
So how is the EDSR,MDSR? would it better than waifu2x on anime-style super-resolution?
Note that I have no knoleged with the deep learning but I would like to proved some idea and discussion here.

(japanese englisgh both ok)

**Comment by nagadomi** [comment]
Several improvements of EDSR have already been implemented in waifu2x. Resudial blocks without BN, L1 loss and Geometric Self-ensemble(we called it TTA). Also waifu2x's resnet_14l model is too slow to use in our use case, probably it is the same issue for EDSR as well. I think those upscaling quality are probably superior, but I guess that it is not much different from resnet_14l.

**Comment by yu45020** [comment]
@2ji3150 
EDSR may not be realistic for CPU only users because it consumes huge amount of memory. I have tried it in PyTorch, and I am not able to scale a 256x256 image with 4G memory. Orz  On the other hand, the authors train the model for around 8 days with multiple GPUs.  

 A deeper model usually improves benchmark scores, but it might not be visually distinguishable for x2 scaling or de-noising.

@nagadomi 
I am reading your source code but fail to find an answer. 
How do you split image into patches and process them ? 

I split a large image into dozens of non-overlapping 128x128 patches, and I dump them into a model. Then I merge all the pieces back to a image. But that image has many square grids.

For example

![11111](https://user-images.githubusercontent.com/28139045/40890396-22700aa6-672a-11e8-9055-63abb1706485.JPG)

**Comment by nagadomi** [comment]
>A deeper model usually improves benchmark scores, but it might not be visually distinguishable for x2 scaling or de-noising.

I guess that is simply overfitting to the downsampling algorithm which is used in benchmark such as bicubic. If you want to develop practical software, you need to improve the average score of various downsampling algorithms.

>I am reading your source code but fail to find an answer.
>How do you split image into patches and process them ?

It's here: https://github.com/nagadomi/waifu2x/blob/c2d00ea0772fbb2ac2ff898389c6773f8369459a/lib/reconstruct.lua#L5

That square grids will occurs when the output is affected by data other than input.
For example, using zero-padding in the convolution layer, it spoils the accuracy in the vicinity of an edge of the output image. Using 3x3 convolution with 1px padding, It can be seen that the edge 1px of the output image becomes unstable. So when using padding, it needs to caluculate unstable area and cut out it.
In waifu2x, padding is not used.

**Comment by yu45020** [comment]
Thanks for the suggestion. I read that part before but didn't realize the problem was caused by padding. 

`I guess that is simply overfitting to the downsampling algorithm which is used in benchmark such as bicubic. If you want to develop practical software, you need to improve the average score of various downsampling algorithms.`

Inspired by VDSR's paper, I train images from various down-sampling method and find deep models perform better than shallow models, which is not surprise. But as the layers stacking up, I am not able to distinguish outputs in x2 scaling.

**Comment by yu45020** [comment]
@2ji3150

> If waifu2x move to Pytorch[or Caffe2,CNTK,Chainer], than we can export ONNX models.

 
[personal ad](https://github.com/yu45020/Waifu2x) I am re-implementing some of waifu2x's models and some other super resolution models in PyTorch. Pre-trained weights can be imported into PyTorch's models without plain.  But the run time (cpu only) is around 5 times longer than the [caffe version](https://github.com/lltcggie/waifu2x-caffe).

**Comment by 2ji3150** [comment]
＠yu45020 
Thank you for telling me. Though I know that the original waifu2x is made for Web service, but I don't really care much about processing speed. Because I want to be able to pursue higher quality on the local PC with mainstream graphics cards(GTX1060 etc).

**Comment by DonaldTsang** [comment]
@yu45020 @2ji3150 how well does the new model perform, I know that some models like Inception is unusable in superesolution (https://github.com/nagadomi/waifu2x/issues/231) but it might be good to start exploring new territory
Examples:
* https://arxiv.org/pdf/1808.03344.pdf (many types)
* https://arxiv.org/ftp/arxiv/papers/1707/1707.05425.pdf (DCSCN and others)
* https://arxiv.org/pdf/1706.09077.pdf (SRCNN and others)

**Comment by 2ji3150** [comment]
Currently I think CUNET do the great job for State of Art on 2D Art Scaling. So I close this issue.

**Comment by tcyrus** [comment]
> If waifu2x moves to PyTorch (or Caffe2, CNTK, Chainer), then we can export ONNX models.

I found a tool that should be able to convert the LuaTorch models to ONNX, but I don't know how well it works ([link](https://github.com/jsenellart/luatorch-onnx-convert)).

PyTorch, Caffe2, CNTK, and Chainer support direct export to ONNX. [onnxmltools](https://github.com/onnx/onnxmltools) supports converting to other frameworks.

Information on making ONNX models can be found [here](https://github.com/onnx/tutorials#converting-to-onnx-format)
